_pylibczi/Image.h

```diff
@@ -0,0 +1,286 @@
+	  std::vector<size_t> m_matrixSizes; // C Y X order or Y X  ( H, W )
+	  libCZI::PixelType m_pixelType;
+	  libCZI::CDimCoordinate m_cdims;
+	  libCZI::IntRect m_xywh;
+	  int m_mIndex;
+
+	  static std::unique_ptr<std::map<libCZI::PixelType, std::string>>
+			  s_pixelToTypeName;
+
+  public:
+	  using ImVec = std::vector<std::shared_ptr<ImageBC> >;
+
+	  ImageBC(std::vector<size_t> shp, libCZI::PixelType pt, const libCZI::CDimCoordinate* cdim,
+			  libCZI::IntRect ir, int mIndex)
+			  :m_matrixSizes(std::move(shp)), m_pixelType(pt), m_cdims(*cdim),
+			   m_xywh(ir), m_mIndex(mIndex) { }
+
+	  size_t calculate_idx(const std::vector<size_t>& idxs);
+
+	  template<typename T>
+	  bool is_type_match();
+
+	  std::vector<size_t> shape() { return m_matrixSizes; }
+
+	  size_t length()
+	  {
+		  return std::accumulate(m_matrixSizes.begin(), m_matrixSizes.end(), (size_t)1,
+				  std::multiplies<>());
+	  }
+
+	  std::vector<std::pair<char, int> > get_valid_indexs(bool isMosaic=false)
+	  {
+		  using CZI_DI = libCZI::DimensionIndex;
+		  std::vector<CZI_DI> sort_order{CZI_DI::S, CZI_DI::T, CZI_DI::C, CZI_DI::Z};
+		  std::vector<std::pair<char, int> > ans;
+		  for (auto di : sort_order) {
+			  int value;
+			  if (m_cdims.TryGetPosition(di, &value)) ans.emplace_back(libCZI::Utils::DimensionToChar(di), value);
+		  }
+		  if(isMosaic) ans.emplace_back('M', m_mIndex);
+		  return ans;
+	  }
+
+	  bool operator<(ImageBC& other)
+	  {
+		  using CZI_DI = libCZI::DimensionIndex;
+		  std::vector<CZI_DI> sort_order{CZI_DI::S, CZI_DI::T, CZI_DI::C, CZI_DI::Z};
+		  for (auto di : sort_order) {
+			  int di_value, other_value;
+			  if (m_cdims.TryGetPosition(di, &di_value) && other.m_cdims.TryGetPosition(di, &other_value) && di_value!=other_value)
+				  return (di_value<other_value);
+		  }
+		  return m_mIndex<other.m_mIndex;
+	  }
+
+	  libCZI::PixelType pixelType() { return m_pixelType; }
+
+	  virtual void load_image(const std::shared_ptr<libCZI::IBitmapData>& pBitmap, size_t channels) = 0;
+
+	  virtual ImVec split_channels(int startFrom) = 0;
+  };
+
+  template<typename T>
+  inline bool
+  ImageBC::is_type_match()
+  {
+	  auto pt = (*s_pixelToTypeName)[m_pixelType];
+	  return (typeid(T).name()==(*s_pixelToTypeName)[m_pixelType]);
+  }
+
+  template<typename T>
+  class Image: public ImageBC {
+	  std::unique_ptr<T[]> m_array;
+
+	  // allow ImageFactory access -> 2 statements below mean ImageFactory is the
+	  // only way to make an image this prevents people from mucking up the order
+	  // indexing of image to memory copying
+	  friend ImageFactory;
+
+	  // private constructor
+
+  public:
+	  // TODO ? make the constructor private ?
+	  /*!
+	   * @brief The Image constructor creates the container and memory for storing an image from a ZeissRaw/CZI file. This class
+	   * is really intended to be created by ImageFactory.
+	   * @param shp The shape of the image can be in a vector or a tuple but <b>MUST be in {C, Y, X} order</b>.
```

Comment:
how hard would it be to fully expand these parameter names? `shp` to `shape`, `pt` to `pixel_type`, etc. I personally would prefer it like that but I don't know C++ standards

Reply:
agree.. for readability, spelling it out is usually preferred.

Reply:
I took a shot at this let me know if there are other points where readability could be improved.